### PR TITLE
stack-trace symbolizer using llvm-symbolizer subprocess

### PR DESCRIPTION
**Pull request by mikea**
When linked it, looks for $LLVM_SYMBOLIZER environment variable and uses it for stacktrace symbolication.

To use it with bazel you should add `--config=fastdbg --test_env=LLVM_SYMBOLIZER=llvm-symbolizer-15`
to your test command.

### CODE REVIEW

File: src/workerd/util/symbolizer.c++ [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+#include <kj/debug.h>
+#include <kj/exception.h>
+#include "dlfcn.h"
+
+// Stack trace symbolizer. To use link this source to the binary.
+// Current implementation shells out to $LLVM_SYMBOLIZER if it is defined.
+
+namespace kj {
+
+struct Subprocess {
```
Comment (kentonv) [inline src/workerd/util/symbolizer.c++:17]: It's easy to imagine KJ growing its own subprocess API someday, which could have symbol name conflicts with the one defined here. So this probably shouldn't be in the KJ namespace. Only `stringifyStackTrace` needs to be declared inside `namespace kj`...

File: src/workerd/util/symbolizer.c++ [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    return status;
+  }
+
+  int pid;
+  int in;
+  int out;
+};
+
+String stringifyStackTrace(ArrayPtr<void* const> trace) {
+  const char* llvmSymbolizer = getenv("LLVM_SYMBOLIZER");
```
Comment (kentonv) [inline src/workerd/util/symbolizer.c++:79]: I tried using this but it appears bazel hides all environment variables from test programs, so it doesn't work unless I run the test directly. What's the right way to make it work when the test runs under bazel?

**Comment by kentonv** [comment]
Can we also update `.bazelrc` so that this produces useful results? Right now I get no line numbers because by default we compile without line number info. Should we add `-gline-tables-only` to the default config?

**Comment by kentonv** [comment]
Hmm, this didn't seem to have any effect in my bazelrc... `build --cxxopt='-gline-tables-only' --host_cxxopt='-gline-tables-only'`

**Comment by mikea** [comment]
bazel's stripping is weird. It worked for me only with `-c dbg`.

https://bazel.build/docs/user-manual#strip

File: src/workerd/util/symbolizer.c++ (hunk 1 above)
Comment (mikea) [reply src/workerd/util/symbolizer.c++:17]: it is in the top comment: `-c dbg --test_env=LLVM_SYMBOLIZER=llvm-symbolizer-15`

Comment (mikea) [reply src/workerd/util/symbolizer.c++:17]: we can always make things default later

File: src/workerd/util/symbolizer.c++ (hunk 2 above)
Comment (mikea) [reply src/workerd/util/symbolizer.c++:79]: (wrong chain)

it is in the top comment: `-c dbg --test_env=LLVM_SYMBOLIZER=llvm-symbolizer-15`

we can always make things default later

**Comment by kentonv** [comment]
`-c dbg` produces prohibitively large binaries for me. `-gline-tables-only` combined with `--strip=never` seems to do the trick.

**Comment by kentonv** [comment]
Here's what I added to `.bazelrc`:

```
# Include basic debug info even in fastbuild mode, otherwise debugging is impossible. Switching
# to debug mode for debugging doesn't really work because debug mode produces insanely large
# binaries.
build:fastbuild --cxxopt='-gline-tables-only' --host_cxxopt='-gline-tables-only'
build:fastbuild --linkopt='-gline-tables-only' --host_linkopt='-gline-tables-only'
build:fastbuild --strip=never
```

File: src/workerd/util/symbolizer.c++ (hunk 2 above)
Comment (kentonv) [reply src/workerd/util/symbolizer.c++:83]: Duh, sorry, I can't read.

**Review by kentonv — APPROVED** [review]
Thanks! I already used this today, it was quite helpful.

File: .bazelrc [hunk 3]
```diff
@@ -28,3 +28,12 @@ build --cxxopt='-Wno-ambiguous-reversed-operator' --host_cxxopt='-Wno-ambiguous-
 build:thin-lto -c opt
 build:thin-lto --cxxopt='-flto=thin'
 build:thin-lto --linkopt='-flto=thin'
+
+# Define a config mode which is fastbuild but with basig debug info.
```
Comment (jasnell) [inline .bazelrc:32]:
```suggestion
# Define a config mode which is fastbuild but with basic debug info.
```

**Review by jasnell — APPROVED** [review]

**Pull request merged**